File: envconfig/vram_budget.go
package envconfig

import (
	"fmt"
	"log/slog"
	"math"
	"strconv"
	"strings"
)

// VRAMBudget is an optional cap on allocatable VRAM (LocalAI LA19).
// Zero value means no cap. Percentage form uses fraction (0,1]; absolute uses bytes.
type VRAMBudget struct {
	fraction float64
	absolute uint64
}

var sizeSuffixes = []struct {
	suffix string
	mult   uint64
}{
	{"KIB", 1 << 10}, {"MIB", 1 << 20}, {"GIB", 1 << 30}, {"TIB", 1 << 40},
	{"KB", 1000}, {"MB", 1000 * 1000}, {"GB", 1000 * 1000 * 1000}, {"TB", 1000 * 1000 * 1000 * 1000},
	{"B", 1},
}

// ParseVRAMBudget accepts "", "80%", "0.8", "12GB", "12GiB", or raw bytes.
func ParseVRAMBudget(s string) (VRAMBudget, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return VRAMBudget{}, nil
	}
	upper := strings.ToUpper(s)
	if strings.HasSuffix(upper, "%") {
		num := strings.TrimSpace(strings.TrimSuffix(upper, "%"))
		v, err := strconv.ParseFloat(num, 64)
		if err != nil {
			return VRAMBudget{}, fmt.Errorf("invalid vram budget percentage %q: %w", s, err)
		}
		return vramBudgetFromFraction(v/100.0, s)
	}
	for _, sfx := range sizeSuffixes {
		if strings.HasSuffix(upper, sfx.suffix) {
			num := strings.TrimSpace(strings.TrimSuffix(upper, sfx.suffix))
			v, err := strconv.ParseFloat(num, 64)
			if err != nil || v < 0 {
				return VRAMBudget{}, fmt.Errorf("invalid vram budget %q", s)
			}
			return VRAMBudget{absolute: uint64(v * float64(sfx.mult))}, nil
		}
	}
	v, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return VRAMBudget{}, fmt.Errorf("invalid vram budget %q", s)
	}
	if v < 0 {
		return VRAMBudget{}, fmt.Errorf("invalid vram budget %q: negative", s)
	}
	if v > 0 && v <= 1 {
		return vramBudgetFromFraction(v, s)
	}
	if v != math.Trunc(v) {
		return VRAMBudget{}, fmt.Errorf("invalid vram budget %q: out of range", s)
	}
	return VRAMBudget{absolute: uint64(v)}, nil
}

func vramBudgetFromFraction(f float64, orig string) (VRAMBudget, error) {
	if f <= 0 {
		return VRAMBudget{}, nil
	}
	if f > 1 {
		return VRAMBudget{}, fmt.Errorf("vram budget %q exceeds 100%%", orig)
	}
	return VRAMBudget{fraction: f}, nil
}

func (b VRAMBudget) IsSet() bool { return b.fraction > 0 || b.absolute > 0 }

func (b VRAMBudget) String() string {
	switch {
	case b.fraction > 0:
		return strconv.FormatFloat(b.fraction*100, 'f', -1, 64) + "%"
	case b.absolute == 0:
		return ""
	case b.absolute%(1<<30) == 0:
		return strconv.FormatUint(b.absolute>>30, 10) + "GiB"
	case b.absolute%(1000*1000*1000) == 0:
		return strconv.FormatUint(b.absolute/1_000_000_000, 10) + "GB"
	default:
		return strconv.FormatUint(b.absolute, 10) + "B"
	}
}

// Ceiling is the byte cap against detectedTotal, clamped so it cannot invent VRAM.
func (b VRAMBudget) Ceiling(detectedTotal uint64) uint64 {
	if !b.IsSet() || detectedTotal == 0 {
		return 0
	}
	var ceil uint64
	if b.fraction > 0 {
		ceil = uint64(float64(detectedTotal) * b.fraction)
	} else {
		ceil = b.absolute
	}
	if ceil > detectedTotal {
		ceil = detectedTotal
	}
	return ceil
}

// Apply caps detected total/free. Unchanged when unset or total is unknown.
func (b VRAMBudget) Apply(detectedTotal, detectedFree uint64) (total, free uint64) {
	if !b.IsSet() || detectedTotal == 0 {
		return detectedTotal, detectedFree
	}
	ceil := b.Ceiling(detectedTotal)
	total = min(detectedTotal, ceil)
	free = min(detectedFree, ceil)
	return total, free
}

// VRAMBudgetFromEnv reads ZEROLLAMA_VRAM_BUDGET. Invalid values are treated as unset.
func VRAMBudgetFromEnv() VRAMBudget {
	s := Var("ZEROLLAMA_VRAM_BUDGET")
	if s == "" {
		return VRAMBudget{}
	}
	b, err := ParseVRAMBudget(s)
	if err != nil {
		slog.Warn("invalid ZEROLLAMA_VRAM_BUDGET, ignored", "value", s, "error", err)
		return VRAMBudget{}
	}
	return b
}

File: runtime/runtime/gpu/vram_budget.py
"""Operator VRAM ceiling (LocalAI LA19 / ZEROLLAMA_VRAM_BUDGET).

Cap is min(detected, budget). Unset leaves probes unchanged. Percentage above
100%% is invalid; absolute values above physical VRAM clamp to detected total.
"""

from __future__ import annotations

import os


def parse_vram_budget(raw: str) -> tuple[float, int]:
    """Return (fraction, absolute_bytes). Both zero means unset.

    fraction in (0, 1]; absolute_bytes when size form. Raises ValueError on
    invalid input. Empty / 0% returns (0, 0).
    """
    s = (raw or "").strip()
    if not s:
        return 0.0, 0
    upper = s.upper()
    if upper.endswith("%"):
        v = float(upper[:-1].strip())
        frac = v / 100.0
        if frac <= 0:
            return 0.0, 0
        if frac > 1:
            raise ValueError(f"vram budget {raw!r} exceeds 100%")
        return frac, 0
    suffixes = (
        ("KIB", 1 << 10),
        ("MIB", 1 << 20),
        ("GIB", 1 << 30),
        ("TIB", 1 << 40),
        ("KB", 1000),
        ("MB", 1000 * 1000),
        ("GB", 1000 * 1000 * 1000),
        ("TB", 1000 * 1000 * 1000 * 1000),
        ("B", 1),
    )
    for suf, mult in suffixes:
        if upper.endswith(suf):
            num = float(upper[: -len(suf)].strip())
            if num < 0:
                raise ValueError(f"invalid vram budget {raw!r}")
            return 0.0, int(num * mult)
    v = float(s)
    if v < 0:
        raise ValueError(f"invalid vram budget {raw!r}: negative")
    if 0 < v <= 1:
        return v, 0
    if v != int(v):
        raise ValueError(f"invalid vram budget {raw!r}: out of range")
    return 0.0, int(v)


def apply_vram_budget(detected_total: int, detected_free: int) -> tuple[int, int]:
    raw = os.environ.get("ZEROLLAMA_VRAM_BUDGET", "").strip()
    if not raw or detected_total <= 0:
        return detected_total, detected_free
    try:
        frac, abs_bytes = parse_vram_budget(raw)
    except ValueError:
        return detected_total, detected_free
    if frac <= 0 and abs_bytes <= 0:
        return detected_total, detected_free
    if frac > 0:
        ceil = int(detected_total * frac)
    else:
        ceil = abs_bytes
    if ceil > detected_total:
        ceil = detected_total
    return min(detected_total, ceil), min(detected_free, ceil)
